File: python/보석쇼핑.py
def solution(gems):
    answer = [0, 0]
    travel = 100000
    start = 0
    end = 100000
    gems_type = []
    for i in range(len(gems)):
        if gems[i] not in gems_type:
            gems_type.append(gems[i])

    # gems_type_tuple = tuple(gems_type)
    # for j in range(len(gems)):
    #     temp = list(gems_type_tuple)
    #     for k in range(j, len(gems)):
    #         if k - j >= travel:
    #             break
    #         if gems[k] in temp:
    #             temp.remove(gems[k])
    #         if temp == []:
    #             start = j
    #             end = k
    #             travel = k - j
    #             break
    # answer[0] = start + 1
    # answer[1] = end + 1

    

    print(answer)
    return answer

solution(["DIA", "RUBY", "RUBY", "DIA", "DIA", "EMERALD", "SAPPHIRE", "DIA"])

File: JavaScript/보석쇼핑.js
function solution(gems) {
    let answer = [1, gems.length];
    const gemTypes = [];
    gems.forEach((g) => {
        if (!gemTypes.includes(g)) gemTypes.push(g);
    });
    const types = gemTypes.length;
    if (types === 1) return [1, 1];
    for (let i = 0; i < gems.length; i++) {
        const gemTypes = [gems[i]];
        let j = i + 1;
        while (j < gems.length) {
            if (!gemTypes.includes(gems[j])) gemTypes.push(gems[j]);
            if (j - i > answer[1] - answer[0] || gemTypes.length === types) break;
            j += 1;
        }
        if (gemTypes.length === types && j - i < answer[1] - answer[0]) answer = [i + 1, j + 1];
    }
    return answer;
}

console.log(solution(["DIA", "RUBY", "RUBY", "DIA", "DIA", "EMERALD", "SAPPHIRE", "DIA"]));